lib/qb_integration/order.rb

```diff
@@ -42,10 +52,91 @@ def cancel
       [200, text]
     end
 
+    def as_flowlink_hash(qbo_order)
+      {
+        id: qbo_order.id,
+        name: qbo_order.doc_number,
+        number: qbo_order.doc_number,
+        created_at: qbo_order.txn_date,
+        line_items: format_line_items(qbo_order.line_items),
+        currency: qbo_order.currency_ref.value,
+        customer: customer_email(qbo_order.customer_ref.value),
+        placed_on: qbo_order.meta_data["create_time"],
+        updated_at: qbo_order.meta_data["last_updated_time"],
+        totals: format_total(qbo_order.line_items),
+        payments: format_payments(qbo_order.payment_ref_number),
+        shipping_address: Address.as_flowlink_hash(qbo_order.ship_address),
+        billing_address: Address.as_flowlink_hash(qbo_order.bill_address)
+      }
+    end
+
     private
 
+    def format_total(line_items)
+      sales_line_details = line_items.select { |line| line.detail_type.to_s == "SalesItemLineDetail" }
+
+      tax_line = sales_line_details.select { |line| line.sales_item_line_detail["item_ref"]["name"].downcase.match(/tax/) }.first
+      shipping_line = sales_line_details.select { |line| line.sales_item_line_detail["item_ref"]["name"].downcase.match(/shipping/) }.first
+      discount_line = sales_line_details.select { |line| line.sales_item_line_detail["item_ref"]["name"].downcase.match(/discount/) }.first
+
+      tax = tax_line && tax_line.amount || BigDecimal("0")
+      shipping = shipping_line && shipping_line.amount || BigDecimal("0")
+      discount = discount_line && discount_line.amount || BigDecimal("0")
+      item = sales_line_details.reduce(0) { |sum, line_details| sum + line_details.amount }.truncate(2).to_s("F").to_f
+
+      {
```

Comment:
cant we move this shipping, discount, tax to be instance methods ? and just leave this hash in this format total method ?

Reply:
It seems this method is doing too much

Reply:
I agree and the class for this is growing a lot bigger than the others, I should start extracting a few parts to their own class